File: typescript/factory_method/transportation/logistics.ts
import { Transport_unit, Truck, Ship, Plane } from "./transport_unit";

export abstract class Logistics {
  public abstract playDelivery(): Transport_unit

  public createTransport(): string {
    return this.playDelivery().deliver()
  }
}

export class RoadLogistics extends Logistics {
  playDelivery(): Transport_unit {
    return new Truck()
  }
}

export class SeaLogistics extends Logistics {
  playDelivery(): Transport_unit {
    return new Ship()
  }
}

export class SkyLogistics extends Logistics {
  playDelivery(): Transport_unit {
    return new Plane()
  }
}

File: dart/creational/factory_method/transportation/bin/logistics.dart
import 'transport_unit.dart';

abstract class Logistics {
  TransportUnit playDelivery();

  String createTransport() {
    return playDelivery().deliver();
  }
}

class RoadLogistics extends Logistics {
  @override
  TransportUnit playDelivery() {
    return Truck();
  }
}

class SeaLogistics extends Logistics {
  @override
  TransportUnit playDelivery() {
    return Ship();
  }
}

class SkyLogistics extends Logistics {
  @override
  TransportUnit playDelivery() {
    return Plane();
  }
}
